File: src/main/java/nl/mpcjanssen/simpletask/util/ListenerList.java
package nl.mpcjanssen.simpletask.util;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

public class ListenerList<L> {
    @NonNull
    private List<L> listenerList = new ArrayList<>();

    public interface FireHandler<L> {
        void fireEvent(L listener);
    }

    public void add(L listener) {
        listenerList.add(listener);
    }

    public void fireEvent(@NonNull FireHandler<L> fireHandler) {
        List<L> copy = new ArrayList<>(listenerList);
        for (L l : copy) {
            fireHandler.fireEvent(l);
        }
    }

    public void remove(L listener) {
        listenerList.remove(listener);
    }

    @NonNull
    public List<L> getListenerList() {
        return listenerList;
    }
}

File: app/src/main/java/nl/mpcjanssen/simpletask/util/ListenerList.kt
package nl.mpcjanssen.simpletask.util

import java.util.*

class ListenerList<L> {
    private val listenerList = ArrayList<L>()

    interface FireHandler<in L> {
        fun fireEvent(listener: L)
    }

    fun add(listener: L) {
        listenerList.add(listener)
    }

    fun fireEvent(fireHandler: FireHandler<L>) {
        val copy = ArrayList(listenerList)
        for (l in copy) {
            fireHandler.fireEvent(l)
        }
    }

    fun remove(listener: L) {
        listenerList.remove(listener)
    }

}
